packages/datastore/src/sync/processors/sync.ts

```
@@ -357,23 +357,39 @@ class SyncProcessor {
 								await Promise.all(promises);
 
 								do {
+									/**
+									 * It's possible that the running processes have been closed
+									 * (e.g. `DataStore.clear` has been called)
+									 */
 									if (!this.runningProcesses.isOpen) {
-										return;
+										logger.error(
+											'No open running processes when starting sync'
+										);
+										return res();
 									}
 
 									const limit = Math.min(
 										maxRecordsToSync - recordsReceived,
 										syncPageSize
 									);
 
-									({ items, nextToken, startedAt } = await this.retrievePage(
-										modelDefinition,
-										lastSync,
-										nextToken,
-										limit,
-										filter,
-										onTerminate
-									));
+									/**
+									 * It's possible that retrievePage will error here, for instance,
+									 * during `authModeRetry` if there are no open running processes.
+									 * An example would be if a model does not exist in the schema.
+									 */
+									try {
+										({ items, nextToken, startedAt } = await this.retrievePage(
+											modelDefinition,
+											lastSync,
+											nextToken,
+											limit,
+											filter,
+											onTerminate
+										));
+									} catch (error) {
+										logger.error('Error retrieving page:', error);
```

For completeness, I think this should `return res()` or `return rej()` here as well.

Looking at the whole loop, I think `return res()` will happen organically with the introduction of the try-catch. But, I'd rather be very explicit.

That said, see my other comment about making sure we do the right thing here. If customers today expect DataStore to fail hard on a broken model, let's reject here and make sure that rejection is "consumed" properly. If they expect all other models to continue functioning, we should log the error and `res()` explicitly.

If I had to guess, today's behavior is a partial failure: My read on it is that DataStore will continue to operate on other models, but will be in a broken state that can't be successfully `stop()`'d or `clear()`'d. If I'm reading that correctly, I think "log the error and continue with other models" is probably the non-breaking change for customers. But, we should confirm that &mdash; maybe you could set up a test app and just add a non-existent model to the client to confirm how the client behaves *today*.

---

Manuel's comment includes excerpts from the spec that address this. TLDR: DataStore should continue merging the rest of the data and make sure to invoke the error handler for non-applicable data. Tested this locally - other than receiving AppSync errors for `retrievePage`, this will continue to fail (for instance, in the event that the customer adds a model to their local schema but doesn't push that change to the backend), due to failures on establishing a subscription. Same applies to adding a non-existent field to a model, etc. In other words, if customers were previously depending on that failure state as a safeguard, it will continue to exist.